### PR TITLE
EBNT-245 introduced Params ABCs for Image, RuntimeEnvironment and RuntimeInstance

### PR DESCRIPTION
**Pull request by i-a-andrianov**
Additionally I was required to eliminate circular import issues which were worked around through lots of indirect references long time ago. Unfortunately (or may be finally and fortunately) without this it was impossible to implement what was expected to be implemented here. For ease of review I split these changes into a separate MR #56 .

### CODE REVIEW

File: src/ebonite/build/runner/docker.py [hunk 1]
```diff
@@ -12,18 +12,10 @@ class DockerRunnerException(Exception):
 
 
 class DockerRuntimeInstance:
```
Comment (mike0sv) [inline src/ebonite/build/runner/docker.py]: what is the difference between this and `DockerContainer`?

Comment (mike0sv) [reply src/ebonite/build/runner/docker.py]: i mean, logically

File: src/ebonite/client/base.py [hunk 2]
```diff
@@ -168,16 +168,15 @@ def run_service(self, name: str, image: Image, environment: RuntimeEnvironment =
         if environment is None:
             environment = self.push_environment(RuntimeEnvironment(env_name))
 
-        params = {k: v for k, v in kwargs.items() if k in {'ports_mapping'}}
+        from ebonite.build import run_docker_img, DockerContainer
```
Comment (mike0sv) [inline src/ebonite/client/base.py]: and again, hardcoded docker (i know we discussed it)
i think a have a vison, how to do this 
we can add abstract method `run_service` to `RuntimeEnvironment.Params` class (and renaming it since it will be more than params container, like wrapper in `Model`) and adding implementation for each child (only docker in our case). then move code from docker helpers there (not sure about it), and in helpers write just `e = DockerEnv(); e.run_service()` or something

File: src/ebonite/client/base.py [hunk 3]
```diff
@@ -168,16 +168,15 @@ def run_service(self, name: str, image: Image, environment: RuntimeEnvironment =
...
+        params = DockerContainer(name, **{k: v for k, v in kwargs.items() if k in {'ports_mapping'}})
```
Comment (mike0sv) [inline src/ebonite/client/base.py]: actually, at first a had another thought - here you hardcode `ports_mapping`, but you can use a function from `pyjackson.utils` which returns class field names (dont remember it's name)

File: src/ebonite/build/runner/docker.py (hunk 1 above)
Comment (i-a-andrianov) [reply src/ebonite/build/runner/docker.py]: `DockerContainer` is a public data model for containers analogical to `DockerImage`. `DockerRuntimeInstance` is an internal class used to encapsulate parameters passed to `DockerRunner`.

File: src/ebonite/client/base.py (hunk 2 above)
Comment (i-a-andrianov) [reply src/ebonite/client/base.py]: Why do we need this if we already have `RunnerBase` ABC for months?

File: tests/repository/metadata/meta_common.py [hunk 4]
```diff
@@ -792,12 +797,16 @@ def test_create_image__saved_image(meta: MetadataRepository, created_image):
 
 def test_update_image__ok(meta: MetadataRepository, created_image):
     author = 'hey'
```
Comment (mike0sv) [inline tests/repository/metadata/meta_common.py]: just to be sure, let's do `author = created_image.author + 'hey'` and same for key

File: src/ebonite/client/base.py (hunk 2 above)
Comment (mike0sv) [reply src/ebonite/client/base.py]: Yes, we had it, but we did not have `RuntimeEnvironment`. But now we have (we did it for persistance), so my proposal is to merge `RunnerBase` logic inside it.

File: src/ebonite/build/runner/docker.py (hunk 1 above)
Comment (mike0sv) [reply src/ebonite/build/runner/docker.py]: this is the same case i think, if we merge `RunnerBase` inside Env, there is no need for two separate "param" classes

Comment (i-a-andrianov) [reply src/ebonite/build/runner/docker.py]: I eliminated `DockerRuntimeInstance` in last version of this PR.

I dislike putting logic into persistence classes as it leads to hacks like `Task.push_model` where low-level object (`Task`) creates a higher level objects (`Ebonite`) to perform requested operation. This is a circular dependency and therefore bad design. Importing classes inside method instead of on top is just a workaround for this but it doesn't make this design any better. In case of this PR I introduced a facade to runners which is used by `Ebonite` for implementation of `run_service`, `is_service_running` and `stop_service`. This way all dependencies are on way only: client => facade => concrete runner => persistence primitives.

File: src/ebonite/client/base.py (hunk 2 above)
Comment (i-a-andrianov) [reply src/ebonite/client/base.py]: Replied above

File: src/ebonite/client/base.py (hunk 3 above)
Comment (i-a-andrianov) [reply src/ebonite/client/base.py]: New design eliminated this part of code, please take a look

File: tests/repository/metadata/meta_common.py (hunk 4 above)
Comment (i-a-andrianov) [reply tests/repository/metadata/meta_common.py]: Could you please describe what do you want to check here by doing that? I'm sorry but I don't get it. This test ensures that `created_image.author` is not 'hey' before update and is 'hey' after. For me it's a guarantee that update works correctly.

Comment (mike0sv) [reply tests/repository/metadata/meta_common.py]: it just a paranoia for this case: someone changes `created_image` fixture to have 'hey' as author and this test will test nothing

Comment (i-a-andrianov) [reply tests/repository/metadata/meta_common.py]: for this we have `assert created_image.params.key != key` before `update_image` call

File: src/ebonite/build/docker.py [hunk 5]
```diff
@@ -70,12 +69,18 @@ def get_uri(self) -> str:
...
         return uri
 
-    def to_core_image(self) -> Image:
-        return Image(self.get_uri(), params=serialize(self))
 
-    @staticmethod
-    def from_core_image(image: Image) -> 'DockerImage':
-        return deserialize(image.params, DockerImage)
+class DockerContainer(RuntimeInstance.Params):
+    # FIXME `ports_mapping` keys could be effectively `int` or `str` but
```
Comment (mike0sv) [inline src/ebonite/build/docker.py]: just added support for int keys t opyjackson, try pyjackson==0.0.25

Comment (i-a-andrianov) [reply src/ebonite/build/docker.py]: ok, hope it helps

File: src/ebonite/client/base.py [hunk 6]
```diff
@@ -153,31 +153,35 @@ def get_environment(self, name: str) -> RuntimeEnvironment:
         """
         return self.meta_repo.get_environment_by_name(name)
 
-    def run_service(self, name: str, image: Image, environment: RuntimeEnvironment = None, **kwargs) -> RuntimeInstance:
+    def run_service(self, name: str, image: Image, environment: RuntimeEnvironment = None,
+                    service_type: str = 'docker', **kwargs) -> RuntimeInstance:
```
Comment (mike0sv) [inline src/ebonite/client/base.py]: why would you need to pass `service_type` if it can be derived from `image` type?
or maybe from env type?

File: src/ebonite/ext/flask/build_templates/flask-site-nginx.conf [hunk 7]
```diff
@@ -1,4 +1,6 @@
 server {
+    listen       9000;
```
Comment (mike0sv) [inline src/ebonite/ext/flask/build_templates/flask-site-nginx.conf]: please check docs and example which maybe affected by this change

File: src/ebonite/build/helpers.py [hunk 8]
```diff
@@ -45,54 +46,58 @@ def build_model_docker(image_params: Union[str, DockerImage], model: Model, serv
...
-    :param container_name: expected name of container to be run
+    :param container_params: expected params (or simply name) of container to be run
     :param image_params: params (or simply name) for docker image to be run
-    :param target_uri: host URI to connect to Docker daemon on, if no given "localhost" is used
+    :param host_params: host params (or simply URI) to connect to Docker daemon on, if no given "localhost" is used
     :param ports_mapping: mapping of exposed ports in container
     :param detach: if `False` block execution until container exits
     :return: nothing
     """
     if isinstance(image_params, str):
```
Comment (mike0sv) [inline src/ebonite/build/helpers.py]: for others you have separate function

File: src/ebonite/build/runner/catalog.py [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+from ebonite.build.runner.docker import DockerRunner
+from ebonite.core.objects import RuntimeInstance
+
+
+class RunnersCatalog:
+    """
+    Catalog for all known :class:`.RunnerBase` implementations
+    """
+
+    _alias_runners: Dict[str, RunnerBase] = {
```
Comment (mike0sv) [inline src/ebonite/build/runner/catalog.py]: now the only option to add new runner is to change ebonite sources. I propose to add a way to dynamically define new Runner implementations - something like Analyzers. and also we can move all docker-related code to 'ebonite.ext.docker' for that matter

File: src/ebonite/build/runner/base.py [hunk 10]
```diff
@@ -1,20 +1,74 @@
...
+        Creates new runtime instance on given name and args
+
+        :param name: name of instance to use
+        :return: created :class:`.RuntimeInstance.Params` subclass instance
+        """
         pass  # pragma: no cover
 
     @abstractmethod
-    def is_running(self, instance) -> bool:
+    def localhost_env(self) -> RuntimeEnvironment.Params:
```
Comment (mike0sv) [inline src/ebonite/build/runner/base.py]: please elaborate on why is this needed

File: src/ebonite/client/base.py [hunk 11]
```diff
@@ -153,31 +153,35 @@ def get_environment(self, name: str) -> RuntimeEnvironment:
...
         :param image: image to run instance from
         :param environment: environment to run instance in, if no given `localhost` is used
+        :param service_type: type of service to be run, default is regular Docker container
         :return: :class:`~ebonite.core.objects.RuntimeInstance` instance representing run instance
         """
-        env_name = 'localhost'
+        from ebonite.build import RunnerBase, RunnersCatalog
+        runner: RunnerBase = RunnersCatalog.get(service_type)
+
+        env_name = f'{service_type}_localhost'
```
Comment (mike0sv) [inline src/ebonite/client/base.py]: doesn't feel right to create and push default env in `run_service` client method. but i dont have a right way to do it in mind yet. i'll write when i got one (if i got one)

File: src/ebonite/build/docker.py (hunk 5 above)
Comment (i-a-andrianov) [reply src/ebonite/build/docker.py]: it did

File: src/ebonite/build/runner/docker.py [hunk 12]
```diff
@@ -11,31 +11,28 @@ class DockerRunnerException(Exception):
...
-                                     instance.params['ports_mapping'])
+    def create_instance(self, name: str, ports_mapping: Dict[int, int] = None, **kwargs) -> DockerContainer:
+        return DockerContainer(name, ports_mapping)
 
+    def localhost_env(self) -> DockerHost:
+        return DockerHost()
 
-class DockerRunner(RunnerBase):
+    def run(self, instance: DockerContainer, image: DockerImage, env: DockerHost, rm=True, detach=True, **kwargs):
+        if not (isinstance(instance, DockerContainer) and isinstance(image, DockerImage) and
```
[the remark this answers is not in the document]
Comment (i-a-andrianov) [reply src/ebonite/build/runner/docker.py]: No, in other methods we have no `image` argument so validation here differs a bit.

File: src/ebonite/client/base.py (hunk 6 above)
Comment (i-a-andrianov) [reply src/ebonite/client/base.py]: It couldn't be derived from `image` type as we could (in future) run Docker images as either Docker container or Kubernetes pod.

It potentially could be derived from `environment` type (if every `instance` type has the only `environment` type and vice versa, and for now I'm not sure about it) but this argument is optional.

In any case we support just Docker for now so there is no need to specify anything really. I suppose to wait for UI/API implementation and support for Kubernetes (or smth else) to discover whether we need to improve usability here and if so how do we need to do it.

File: src/ebonite/build/helpers.py (hunk 8 above)
Comment (i-a-andrianov) [reply src/ebonite/build/helpers.py]: What do you mean by "others" here?

File: src/ebonite/build/runner/base.py (hunk 10 above)
Comment (i-a-andrianov) [reply src/ebonite/build/runner/base.py]: This method is called by `Ebonite.run_service` to create localhost `Environment` if no explicit `Environment` given. For example, our testing code has never specified any environments. This solution dramatically simplifies things in testing / REPL scenarios. For me this has more usefulness than effort to implement.

File: src/ebonite/client/base.py (hunk 11 above)
Comment (i-a-andrianov) [reply src/ebonite/client/base.py]: For me it's OK but both of us are obviously subjective. My motivation to do this is described above.

File: src/ebonite/ext/flask/build_templates/flask-site-nginx.conf (hunk 7 above)
Comment (i-a-andrianov) [reply src/ebonite/ext/flask/build_templates/flask-site-nginx.conf]: I didn't find anything. I couldn't be sure but it looks reasonable as we always used port 9000 for all servers and it is even hardcoded in the client as default port.

File: src/ebonite/build/runner/catalog.py (hunk 9 above)
Comment (i-a-andrianov) [reply src/ebonite/build/runner/catalog.py]: This looks like a reasonable idea. But it could take a fair amount of time and therefore should be implemented if we plan to have other runner implementations soon. I'm not Product Owner but for now I don't think this is true.

File: src/ebonite/build/runner/docker.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (mike0sv) [reply src/ebonite/build/runner/docker.py]: i see, but this inconsistency ugnetaet me a little bit

File: src/ebonite/ext/flask/build_templates/flask-site-nginx.conf (hunk 7 above)
Comment (mike0sv) [reply src/ebonite/ext/flask/build_templates/flask-site-nginx.conf]: this is internal port in container, so maybe there are wrong port mappings. full search for string "80" in source would reveal that, i think

File: src/ebonite/build/helpers.py (hunk 8 above)
Comment (mike0sv) [reply src/ebonite/build/helpers.py]: for host and instance params

File: src/ebonite/build/runner/catalog.py (hunk 9 above)
Comment (mike0sv) [reply src/ebonite/build/runner/catalog.py]: this is important not because we are planning to add runner implementations, but because we must provide a way for others to add their custom runners. i agree this could be another PR, but no need to postpone that until the end of times)

File: src/ebonite/build/runner/base.py (hunk 10 above)
Comment (mike0sv) [reply src/ebonite/build/runner/base.py]: i agree that things must be simplified, but my intuition tells me there is a better way. give me some time to think about it, if i wont come up with something, we'll do it your way

File: src/ebonite/ext/flask/build_templates/flask-site-nginx.conf (hunk 7 above)
Comment (i-a-andrianov) [reply src/ebonite/ext/flask/build_templates/flask-site-nginx.conf]: "wrong port mappings" - wrong for what? I want you to remember that our servers doesn't encapsulate port number. Therefore it's important to employ the same port for all servers. Otherwise both examples and `DockerContainer.__init__` which specifies `{9000: 9000}` as default mapping will work or break depending on a server being in charge. Flask and asyncio use 9000, client - 9000, nginx Flask - 80 (which is nginx default). That's the reason for change here.

I performed search and found that Jupyter example used nginx Flask and relied on port 80. Fixed it. Now we could change server implementation in this tutorial without touching ports.

File: src/ebonite/build/helpers.py (hunk 8 above)
Comment (i-a-andrianov) [reply src/ebonite/build/helpers.py]: "Others" are used in multiple methods and therefore were extracted to helpers to avoid duplication. Why would I extract 2 lines of code to separate method if it is used just once?

File: src/ebonite/build/runner/docker.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (i-a-andrianov) [reply src/ebonite/build/runner/docker.py]: sad to be you :man_shrugging:

File: src/ebonite/ext/flask/build_templates/flask-site-nginx.conf (hunk 7 above)
Comment (mike0sv) [reply src/ebonite/ext/flask/build_templates/flask-site-nginx.conf]: i meant they still use 80 instead of 9000. no all is ok
